src/main/java/io/confluent/connect/jdbc/JdbcSinkConnector.java

```diff
@@ -65,8 +67,20 @@ public ConfigDef config() {
 
   @Override
   public Config validate(Map<String, String> connectorConfigs) {
-    // TODO cross-fields validation here: pkFields against the pkMode
-    return super.validate(connectorConfigs);
+    /** get configuration parsed and validated individually */
+    Config config = super.validate(connectorConfigs);
+
+    config.configValues()
+        .stream()
+        /** use recommended values for configuration validation
+         * {@link PrimaryKeyModeRecommender#validValues}
+         */
+        .filter(cfg -> PK_MODE.equals(cfg.name())
+            && !cfg.recommendedValues().contains(cfg.value()))
+        .findFirst()
+        .ifPresent(cfg -> cfg.addErrorMessage("'" + cfg.value() + "' is not valid"));
```

Comment:
This is a really interesting, lightweight approach that allows us to implement the cross-property logic for acceptable primary key modes in a single place and reuse it for both programmatic UIs (which leverage the "recommended values" for a property) and now preflight validation.

The one thing that makes me a little hesitant is that the error message we're giving to the user isn't very informative. We could say something like "please only use a recommended value for this property", but for users querying the REST API directly without some kind of programmatic UI in front, they might not even know how to retrieve that information (by hitting the `PUT /connector-plugins/{connectorType}/config/validate` endpoint), and even if they do, it'd probably be a little painful to have to make that extra request instead of being given the relevant information directly in the error message.

Do you think there's a way we can isolate the shared logic for validation and our implementation of the `Recommender` interface for this property in a way that allows us to define the relationship between the `pk.mode` and `delete.enabled` properties once, but also generate user-actionable error messages during validation? I haven't thought too much about how we might structure this but am happy to talk here or offline if you'd like to iterate on possibilities.

Reply:
the error message is changed to ```
'pk.mode' can only have value in list [record_key]
```

> Do you think there's a way we can isolate the shared logic for validation and our implementation of the Recommender interface for this property in a way that allows us to define the relationship between the pk.mode and delete.enabled properties once, but also generate user-actionable error messages during validation? 

that is great idea. I am thinking maybe something like ```

interface CrossFieldsValidator extends ConfigDef.Recommender {
 // some method to report more sensible error message
}
```

I need give it more thought, but can we not block this PR before we can settle on a solution?

Reply:
the reason I want the error message to be as generic as possible is I'm thinking we may extend validation on other properties as well.

Reply:
The changes you made are definitely an improvement, but one thing that gives me pause is that we're still ultimately giving users a less-informative error message than the one that they currently see when there's an issue with the combination of `pk.mode`/`delete.mode` that they've used: https://github.com/confluentinc/kafka-connect-jdbc/blob/9811eec6a876ff3541cd415d83672cd527094abf/src/main/java/io/confluent/connect/jdbc/sink/BufferedRecords.java#L333

This message provides all of the information that would be given in the current PR, plus it explains the relationship between the `pk.mode` property and the `delete.enabled` property. With the current approach, users might see that the only `pk.mode` that they could configure the connector with is `record_key`, but they wouldn't know why, or that they could also disable deletes by setting `delete.enabled` to `false` instead of changing the `pk.mode`.

I don't know if there really is an easy way to codify these types of relationships in a generic, reusable fashion. So then the question becomes one of whether it's worth it to retain clear error messages at the cost of higher complexity in the code base. I think I'd prefer clearer error messages but would be interested in thoughts from you and others.

FWIW, we added preflight validation for the BigQuery connector recently and you can see an example of the pattern we used for that (although nothing included leveraging recommended values for validation) with these snippets:
- https://github.com/confluentinc/kafka-connect-bigquery/blob/b090f8566ce0cae74795b49337f6d804747e2af7/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkConnector.java#L55-L65
- https://github.com/confluentinc/kafka-connect-bigquery/blob/b090f8566ce0cae74795b49337f6d804747e2af7/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java#L737-L776
- https://github.com/confluentinc/kafka-connect-bigquery/blob/b090f8566ce0cae74795b49337f6d804747e2af7/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/MultiPropertyValidator.java
- https://github.com/confluentinc/kafka-connect-bigquery/blob/b090f8566ce0cae74795b49337f6d804747e2af7/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/PartitioningTypeValidator.java

(Funnily enough, we probably should have actually added `Recommender` logic for this exact example since it's enforcing a very similar relationship to the one in this PR: if your connector config meets some specific condition (in that case, automatic table creation is enabled and the connector should write to tables using BigQuery decorator syntax), then the set of permitted values for a specific property (in this case, the `timePartitioningType`) are restricted, and if that condition is not met, then the set of permitted values is not restricted. But the end result that all values are recommended at the moment, but you get a clear error message when you use one that is not permitted with the current connector config indicating what's wrong, why, and what to change, still provides a decent UX overall.)

Reply:
> The changes you made are definitely an improvement, but one thing that gives me pause is that we're still ultimately giving users a less-informative error message than the one that they currently see when there's an issue with the combination of `pk.mode`/`delete.mode` that they've used:
> 
> https://github.com/confluentinc/kafka-connect-jdbc/blob/9811eec6a876ff3541cd415d83672cd527094abf/src/main/java/io/confluent/connect/jdbc/sink/BufferedRecords.java#L333

Thank you for the link, I missed existing error message for this particular error.

Reply:
> I don't know if there really is an easy way to codify these types of relationships in a generic, reusable fashion. So then the question becomes one of whether it's worth it to retain clear error messages at the cost of higher complexity in the code base. I think I'd prefer clearer error messages but would be interested in thoughts from you and others.

I agree, there is some balancing need be kept. 

General speaking, I prefer simple/reusable code logic even the cost is less clear error messages, which I believe can be made up by better documentation and run-books. 

I believe your idea is sound to combine recommendation and cross-fields validation into an reusable abstraction. We can explore it further for sure.

Thank you again for spending time and giving a lot of thought on this PR.